src/compiler/gcc.rs

```diff
@@ -75,142 +76,188 @@ impl CCompilerImpl for GCC {
-    let mut multiple_input = false;
     let mut split_dwarf = false;
     let mut need_explicit_dep_target = false;
 
     // Custom iterator to expand `@` arguments which stand for reading a file
     // and interpreting it as a list of more arguments.
-    let mut it = ExpandIncludeFile {
+    let it = ExpandIncludeFile {
         stack: arguments.iter().rev().map(|a| a.to_owned()).collect(),
         cwd: cwd,
     };
-    while let Some(arg) = it.next() {
-        if let Some(s) = arg.to_str() {
-            let mut handled = true;
-            match s {
-                "-c" => compilation = true,
-                "-o" => output_arg = it.next(),
-                "-gsplit-dwarf" => {
-                    split_dwarf = true;
-                    common_args.push(arg.clone());
-                }
-                // Arguments that take a value.
-                // -MF and -MQ are in this set but are handled separately
-                // because they are also preprocessor options.
-                a if argument_takes_value(a) => {
-                    let args = if a == "-include" {
-                        &mut preprocessor_args
-                    } else {
-                        &mut common_args
-                    };
-                    args.push(arg.clone());
-                    if let Some(arg_val) = it.next() {
-                        args.push(arg_val);
-                    }
-                },
-                "-MF" |
-                "-MQ" => {
-                    preprocessor_args.push(arg.clone());
-                    if let Some(arg_val) = it.next() {
-                        preprocessor_args.push(arg_val);
+
+    for item in ArgsIter::new(it, arg_info) {
+        // Refuse to cache arguments such as "-include@foo" because they're a
+        // mess. See https://github.com/mozilla/sccache/issues/150#issuecomment-318586953
+        if let Argument::WithValue(_, ref v, CanBeSeparated(_)) = item.arg {
+            if OsString::from(v.clone()).to_string_lossy().starts_with("@") {
+                return CompilerArguments::CannotCache("@");
+            }
+        }
+        match item.data {
+            Some(TooHard) => {
+                return CompilerArguments::CannotCache(item.arg.to_str().expect(
+                    "Can't be Argument::Raw/UnknownFlag",
+                ))
+            }
+            Some(SplitDwarf) => split_dwarf = true,
+            Some(DoCompilation) => compilation = true,
+            Some(Output) => output_arg = item.arg.get_value().map(|s| s.unwrap_path()),
+            Some(NeedDepTarget) => need_explicit_dep_target = true,
+            Some(DepTarget) => dep_target = item.arg.get_value().map(OsString::from),
+            Some(PreprocessorArgument) |
+            Some(PassThrough) |
+            Some(Language) => {}
+            None => {
+                match item.arg {
+                    Argument::Raw(ref val) => {
+                        if input_arg.is_some() {
```

Comment:
Doesn't this regress your change to make linking invocations produce `NotCompilation`?